includes/class-sensei-lesson.php

```diff
@@ -3822,6 +3822,53 @@ public function get_course_id( $lesson_id ) {
 
 	}
 
+	/**
+	 * Returns the course ID for a given set of lessons (with the key being the lesson ID)
+	 *
+	 * @since $$next-version$$
+	 * @access public
+	 *
+	 * @param array<int> $lesson_ids Array of Lesson IDs.
+	 * @return array<int, int|false> Array where the key is the lesson ID, and the value is the course ID (or false if not found)
+	 */
+	public function get_course_ids( $lesson_ids ) {
+		global $wpdb;
+
+		sort( $lesson_ids, SORT_NUMERIC );
+
+		$cache_key     = 'lesson/get-course-ids/' . implode( ',', $lesson_ids );
+		$cache_group   = 'sensei';
+		$cached_result = wp_cache_get( $cache_key, $cache_group );
+		if ( false !== $cached_result ) {
+			return $cached_result;
+		}
+		$courses_by_lesson = array_fill_keys( $lesson_ids, false );
+		$placeholders      = implode( ', ', array_fill( 0, count( $lesson_ids ), '%d' ) );
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare
+		$query = $wpdb->prepare(
+			"
+				SELECT lesson.ID AS lesson_id, course.ID AS course_id FROM {$wpdb->posts} lesson
```

Comment:
Hum! I think @alexsanford is right! Probably it's not the case for `posts` vs `postmeta`. But I can check it later to confirm since I'll work there today.

p1651851724334379-slack-C07418EJ0

Reply:
Yep, as @alexsanford explained, the reasoning for this join is to check if the course iD is valid and refers to a course post type, without having to check again in the code. You can verify those checks on the `get_course_id` method, where that's checked individually.

I'm thinking here about the possibilities related to this, @renatho, as I know that you fixed a bug related to a similar JOIN on Learnomattic in the past, however, if I change this to avoid JOINs, we will basically do the same as a `get_course_id` in a loop (because we need to check if both course ID and lesson ID is valid), so...yeah.

Reply:
@renatho I just checked your slack thread, and as I suspected, the issue is with the `wp_users` table. That one is "special" on a multisite installation like dotcom, because all users across all sites are stored in the same table. It's probably quite large, and I can understand why it might need to be handled differently.

I imagine the reason that y'all couldn't find reference to the `wp_*_comments` table (I also couldn't find `wp_*_posts` or `wp_*_postmeta` for that matter) is because these tables are created (with a unique prefix) for each new site.

So I still think it's safe to assume that those tables should be safe to `JOIN`, even on dotcom, but would be nice to test/verify that 😅

Reply:
I decided to ping Igor to check that. Here's the thread: p1656959044763259-slack-C07418EJ0

Reply:
Hey! Sorry for the delay in answering that, but I just checked it, and it works there. Sorry for the noise here!

Reply:
>I think part of the issue here is validation. If we can JOIN then we can check that the Course ID is valid, without the need to loop through them all and look for that course in the DB (which is part of the reason for the original perf issue).

@alexsanford I agree, but... what is it really validating? I mean, I can't imagine any "normal" flow that could make these values be something "different" from a course_id. And anyway, if it happened it could be that, by coincidence the "broken" id matched a "valid" course still... So... I am not sure these validations are adding much.
Maybe we want to ensure that courses are not "deleted" but still we are not validating that, as it could be that the status for the courses was "deleted".

Anyway I always prefered to avoid joins because some times are not really necessary and just overcomplicate queries. But I am ok with continuing with it if you all agree on it's need.

Reply:
> but... what is it really validating?

Unexpected weirdness that could break stuff 😅 

It may be unlikely for the validation to fail, but personally I think it's still worth doing. Never underestimate the ability for a WP database to get messed up haha.

> Maybe we want to ensure that courses are not "deleted" but still we are not validating that

That's a good point. CC @fjorgemota do you think we should add this status check? I'm not sure this is checked in `get_course_id` either, so maybe it's ok to leave this out?